pcdsdevices/ccm.py

```diff
@@ -0,0 +1,185 @@
+import time
+
+import numpy as np
+from ophyd.device import Component as Cpt, FormattedComponent as FCpt
+from ophyd.pseudopos import PseudoPositioner, PseudoSingle
+from ophyd.pv_positioner import PVPositionerPC
+from ophyd.signal import EpicsSignal, EpicsSignalRO, AttributeSignal
+
+from .epics_motor import IMS
+from .inout import InOutPositioner
+from .pseudopos import SyncAxes
+
+
+# Default constants copied verbatim from old python
+gTheta0 = 14.9792 * np.pi/180
+gSi111dspacing = 3.1356011499587773
+gSi511dspacing = 1.0452003833195924
+gdspacing = gSi111dspacing
+gR = 3.175
+gD = 231.303
+
+
+# Calculations between alio position and energy, with all intermediates.
+def theta_to_alio(theta, gTheta0, gR, gD):
+    """
+    Converts theta angle (rad) to alio position (mm)
+    """
+    return gR * (1/np.cos(theta)-1) + gD * np.tan(theta - gTheta0)
+
+
+def alio_to_theta(alio, gTheta0, gR, gD):
+    """
+    Converts alio position (mm) to theta angle (rad)
+    """
+    low = -1.0
+    high = 1.0
+    timeout = 1.0
+    start = time.time()
+    while time.time() - start < timeout:
+        theta_guess = (low+high)/2
+        alio_calc = theta_to_alio(theta_guess, gTheta0, gR, gD)
+        if np.isclose(alio, alio_calc):
+            break
+        elif alio_calc > alio:
+            high = theta_guess
+        else:
+            low = theta_guess
```

Comment:
Oh boy. This is unclean. If you tell me its necessary than its neccesary I suppose. Should `np.isclose` have a set-able tolerance?

Reply:
You are free to find the analytic inversion of the other method if you'd like. I think this binary search inverter should be a generic function in a utils package. There is no need to have a settable tolerance because nobody will ever use the knob. I might tweak the default tolerance to get a tighter fit.